Add pages/recipes for common questions (#3106)
cerbos/cerbos: docs/modules/recipes/pages/jwt-claims.adoc

include::ROOT:partial$attributes.adoc[]

= Working with JWT claims in policies

include::ROOT:partial$version-check.adoc[]

> When xref:configuration:auxdata.adoc[auxiliary data] is configured, Cerbos can verify and decode JWTs sent with requests. The decoded claims are available in policy conditions as `request.aux_data.jwt`. This page covers common patterns for using JWT claims in xref:policies:conditions.adoc[conditions].

Reviewer: ```suggestion
When xref:configuration:auxdata.adoc[auxiliary data] is configured, Cerbos can verify and decode JWTs sent with requests. The decoded claims are available in policy conditions as `request.auxData.jwt`. This page covers common patterns for using JWT claims in xref:policies:conditions.adoc[conditions].
```

== Accessing claims

> Claims are accessed using dot notation from the `request.aux_data.jwt` object.

Reviewer: ```suggestion
Claims are accessed using dot notation from the `request.auxData.jwt` object.
```

.Simple string claim
[source,yaml,linenums]
----
condition:
  match:

>     expr: request.aux_data.jwt.iss == "https://auth.example.com"

Reviewer: ```suggestion
    expr: request.auxData.jwt.iss == "https://auth.example.com"
```

----

.Nested claim
[source,yaml,linenums]
----
condition:
  match:

>     expr: request.aux_data.jwt.custom.department == "engineering"

Reviewer: ```suggestion
    expr: request.auxData.jwt.custom.department == "engineering"
```

----

> TIP: Use the `has()` macro to safely check for the existence of optional claims before accessing them: `has(request.aux_data.jwt.custom) && request.aux_data.jwt.custom.role == "admin"`.

Reviewer: ```suggestion
TIP: Use the `has()` macro to safely check for the existence of optional claims before accessing them: `has(request.auxData.jwt.custom) && request.auxData.jwt.custom.role == "admin"`.
```

== Array claims

JWT claims like `roles`, `groups`, or `scope` are often arrays or space-delimited strings.

.Check if a value exists in an array claim
[source,yaml,linenums]
----
condition:
  match:
    expr: >

>       "admin" in request.aux_data.jwt.roles

Reviewer: ```suggestion
      "admin" in request.auxData.jwt.roles
```

----

.Check if any value from one list exists in another
[source,yaml,linenums]
----
condition:
  match:
    expr: >

>       request.aux_data.jwt.groups.exists(g, g in ["engineering", "platform"])

Reviewer: ```suggestion
      request.auxData.jwt.groups.exists(g, g in ["engineering", "platform"])
```

----

.Space-delimited scope strings (common in OAuth2)
[source,yaml,linenums]
----
condition:
  match:
    expr: >

>       "read:documents" in request.aux_data.jwt.scope.split(" ")

Reviewer: ```suggestion
      "read:documents" in request.auxData.jwt.scope.split(" ")
```

----

== Numeric claims

JWT numeric values may arrive as different types depending on the token issuer. Use explicit type conversion for safe comparisons:

[source,yaml,linenums]
----
condition:
  match:
    expr: >

>       int(request.aux_data.jwt.org_id) == int(R.attr.org_id)

Reviewer: ```suggestion
      int(request.auxData.jwt.org_id) == int(R.attr.org_id)
```

----

== Using JWT claims in derived roles

A common pattern is to create xref:policies:derived_roles.adoc[derived roles] based on JWT claims.

.Grant an `internal_user` derived role based on a custom claim
[source,yaml,linenums]
----
apiVersion: "api.cerbos.dev/v1"
derivedRoles:
  name: jwt_roles
  definitions:
    - name: internal_user
      parentRoles: ["user"]
      condition:
        match:
          expr: >

>             has(request.aux_data.jwt.is_internal) && request.aux_data.jwt.is_internal == true

Reviewer: ```suggestion
            has(request.aux_data.jwt.is_internal) && request.auxData.jwt.is_internal == true
```
